geode-core/src/test/java/org/apache/geode/internal/cache/ha/HARegionQueueTest.java

```diff
@@ -183,4 +203,347 @@ public void isQueueInitializedWithWaitReturnsFalseIfNotInitializedAfterWait() th
+    spy.hasSynchronizedWithPrimary.set(true);
+
+    spy.synchronizeQueueWithPrimary(primary, internalCache);
+    verify(spy, never()).runSynchronizationWithPrimary(primary, internalCache);
+  }
+
+  @Test
+  public void doNotRunSynchronizationWithPrimaryIfSynchronizationIsInProgress() {
+    HARegionQueue spy = spy(haRegionQueue);
+    spy.synchronizeWithPrimaryInProgress.set(true);
+
+    spy.synchronizeQueueWithPrimary(primary, internalCache);
+    verify(spy, never()).runSynchronizationWithPrimary(primary, internalCache);
+  }
+
+  @Test
+  public void doNotRunSynchronizationWithPrimaryIfGIINotFinished() {
+    HARegionQueue spy = spy(haRegionQueue);
+
+    spy.synchronizeQueueWithPrimary(primary, internalCache);
+    verify(spy, never()).runSynchronizationWithPrimary(primary, internalCache);
+  }
+
+  @Test
+  public void doNotRunSynchronizationWithPrimaryIfPrimaryHasOlderThanGEODE_1_14_0Version() {
+    HARegionQueue spy = spy(haRegionQueue);
+    spy.doneGIIQueueing.set(true);
+    when(primary.getVersionOrdinal()).thenReturn((short) (KnownVersion.GEODE_1_14_0.ordinal() - 1));
+
+    spy.synchronizeQueueWithPrimary(primary, internalCache);
+    verify(spy, never()).runSynchronizationWithPrimary(primary, internalCache);
+  }
+
+  @Test
+  public void runSynchronizationWithPrimaryIfPrimaryIsGEODE_1_14_0Version() {
+    HARegionQueue spy = spy(haRegionQueue);
+    spy.doneGIIQueueing.set(true);
+    when(primary.getVersionOrdinal()).thenReturn(KnownVersion.GEODE_1_14_0.ordinal());
+    doNothing().when(spy).runSynchronizationWithPrimary(primary, internalCache);
+
+    spy.synchronizeQueueWithPrimary(primary, internalCache);
+    verify(spy).runSynchronizationWithPrimary(primary, internalCache);
+  }
+
+  @Test
+  public void runSynchronizationWithPrimaryIfPrimaryIsLaterThanGEODE_1_14_0Version() {
+    HARegionQueue spy = spy(haRegionQueue);
+    spy.doneGIIQueueing.set(true);
+    when(primary.getVersionOrdinal()).thenReturn((short) (KnownVersion.GEODE_1_14_0.ordinal() + 1));
+    doNothing().when(spy).runSynchronizationWithPrimary(primary, internalCache);
+
+    spy.synchronizeQueueWithPrimary(primary, internalCache);
+    verify(spy).runSynchronizationWithPrimary(primary, internalCache);
+  }
+
+  @Test
+  public void getGIIEventsReturnsCorrectEvents() {
+    HARegionQueue spy = spy(haRegionQueue);
+    List<EventID> giiEvents;
+    spy.positionBeforeGII = 1;
+    spy.positionAfterGII = 4;
+    HAEventWrapper wrapper1 = mock(HAEventWrapper.class);
+    HAEventWrapper wrapper2 = mock(HAEventWrapper.class);
+    when(wrapper1.getEventId()).thenReturn(id1);
+    when(wrapper2.getEventId()).thenReturn(id2);
+    Region.Entry entry1 = mock(Region.Entry.class);
+    Region.Entry entry2 = mock(Region.Entry.class);
+    Region.Entry entry3 = mock(Region.Entry.class);
+    when(entry1.getValue()).thenReturn(wrapper1);
+    when(entry3.getValue()).thenReturn(wrapper2);
+    when(entry2.getValue()).thenReturn(id3);
```

Comment:
The numbering of things is a little confusing here. Is there a reason that `entry3` returns `wrapper2`  and `entry2` returns `id3`?

Reply:
No particular reason, just want to test non HAEventWrapper in an entry case. To be clear, I change the id3 to an object.

Reply:
Thanks for all the comments and change recommendations.